### PR TITLE
Implement overhaul of cause system

### PR DESCRIPTION
**Pull request by Deamon5550**
[SpongeApi](https://github.com/SpongePowered/SpongeAPI/pull/1370) | **SpongeCommon** | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1670) | [SpongeVanilla](https://github.com/SpongePowered/SpongeVanilla/pull/299)

See api PR for description

Current causes

Event | Bleeding | Now
-------|-----------|-------
GrantAchievementEvent | `Cause[{Name=Source, Object={EntityPlayerMP['Deamon_t'/293, l='world', x=-185.83, y=72.00, z=339.17]}]` | `Cause (size 1): [EntityPlayerMP['Deamon_t'/293, l='world', x=-185.83, y=72.00, z=339.17]] Context (0): []`

### CODE REVIEW

**Comment by ryantheleach** [comment]
I notice that you have a table in the original post, do you need help in dumping all the current causes ?

Or how can I assist with testing?

**Comment by Deamon5550** [comment]
@ryantheleach yeah if you want to go through and dump current causes for bleeding that would be a great help

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityLivingBase.java [hunk 1]
```diff
@@ -256,73 +256,79 @@ public void onDeath(DamageSource cause) {
...
+        // Sponge Start - Don't send the state if this is a human. Fixes ghost items on client.
+        if (!((EntityLivingBase) (Object) this instanceof EntityHuman)) {
+            this.world.setEntityState((EntityLivingBase) (Object) this, (byte) 3);
+        }
+        if (causeTracker != null && this.tracksEntityDeaths && !properlyOverridesOnDeathForCauseTrackerCompletion()) {
+            this.tracksEntityDeaths = false;
+            causeTracker.completePhase(EntityPhase.State.DEATH);
+        }
+
+        if (frame != null) {
```
Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityLivingBase.java]: Do these really pop up as often as I think they do? Frame.closeOnMainThread(frame)?

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityLivingBase.java [hunk 2]
```diff
@@ -256,73 +256,79 @@ public void onDeath(DamageSource cause) {
...
-            EntityLivingBase entitylivingbase = this.getAttackingEntity();
-    
-            if (this.scoreValue >= 0 && entitylivingbase != null) {
-                entitylivingbase.awardKillScore((EntityLivingBase) (Object) this, this.scoreValue, cause);
-            }
-    
-            if (entity != null) {
-                entity.onKillEntity((EntityLivingBase) (Object) this);
+        final boolean isMainThread = !this.world.isRemote || Sponge.isServerAvailable() && Sponge.getServer().isMainThread();
+        final CauseStackFrame frame = isMainThread ? Sponge.getCauseStackManager().pushCauseFrame() : null;
```
Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityLivingBase.java]: Sponge.getCauseStackManager().pushFrameIfMain(); ?

File: src/main/java/org/spongepowered/common/event/tracking/phase/tick/BlockTickPhaseState.java [hunk 3]
```diff
@@ -78,59 +71,37 @@ LocatableBlock getLocatableBlockSourceFromContext(PhaseContext context) {
...
-                    final SpawnEntityEvent
-                            spawnEntityEvent =
-                            SpongeEventFactory.createSpawnEntityEvent(cause, capturedEntities);
-                    SpongeImpl.postEvent(spawnEntityEvent);
-                    for (Entity entity : spawnEntityEvent.getEntities()) {
-                        if (entityCreator != null) {
-                            EntityUtil.toMixin(entity).setCreator(entityCreator.getUniqueId());
+        try (CauseStackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+            Sponge.getCauseStackManager().pushCause(locatableBlock);
+            Sponge.getCauseStackManager().addContext(EventContextKeys.SPAWN_TYPE, InternalSpawnTypes.DROPPED_ITEM);
```
Comment (Faithcaio) [inline src/main/java/org/spongepowered/common/event/tracking/phase/tick/BlockTickPhaseState.java]: This adds a dropped item context to a lot of ChangeBlockEvents.
.Break (always in creative when no items drop / fire)
.Modify (tested with fire)

**Comment by Faithcaio** [comment]
Things to look into:

# ChangeBlockEvent

 - [ ] Indirect ChangeBlockEvents by Gravity. (Sand falling because you broke the block below)
   - Player is only the owner (probably because I placed the sand). Notifier and/or cause is missing.
- [ ] Redstone activation:
   -  Wire is the cause? Instead of a redstone source.  
<details><summary>Setup and Cause printout</summary><p>

### Setup

  image

### CausePrintout
```
ChangeBlockEvent$Place$Impl
Tra: @-72:392 Block{minecraft:air} -> Block{minecraft:redstone_torch}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
C|1EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]


ChangeBlockEvent$Modify$Impl
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-75:392 Block{minecraft:redstone_lamp} -> Block{minecraft:lit_redstone_lamp}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
C|1SpongeLocatableBlock{blockState=minecraft:redstone_wire[east=none,north=none,power=14,south=none,west=none], position=(-74, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@75f431dd}
C|2EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
```


# With repeater: 
image


```
ChangeBlockEvent$Place$Impl
Tra: @-70:392 Block{minecraft:air} -> Block{minecraft:redstone_torch}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]
C|1EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]


ChangeBlockEvent$Modify$Impl
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]
C|1SpongeLocatableBlock{blockState=minecraft:redstone_torch[facing=up], position=(-70, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@853c3f2}
C|2EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]


ChangeBlockEvent$Modify$Impl
Tra: @-72:392 Block{minecraft:unpowered_repeater} -> Block{minecraft:powered_repeater}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_lamp} -> Block{minecraft:lit_redstone_lamp}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-68,79, y=4,00, z=394,89]
C|1SpongeLocatableBlock{blockState=minecraft:unpowered_repeater[delay=1,facing=east,locked=false], position=(-72, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@4b1b978f}
C|2WorldServer{LevelName=world, DimensionId=0, DimensionType=minecraft:overworld}
```


</p></details>

 - [ ] Redstone deactivation
    - Again Redstone wire is part of the cause
    - Also the same block has lots of transactions in one Event
<details><summary>Setup and Cause printout</summary><p>

   - same setup
### CausePrintout
```
ChangeBlockEvent$Break$Impl
Tra: @-72:392 Block{minecraft:redstone_torch} -> Block{minecraft:air}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
C|1EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]


ChangeBlockEvent$Modify$Impl
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-74:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
C|1SpongeLocatableBlock{blockState=minecraft:redstone_wire[east=none,north=none,power=0,south=none,west=none], position=(-74, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@63d24cd0}
C|2EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]


ChangeBlockEvent$Modify$Impl
Tra: @-75:392 Block{minecraft:lit_redstone_lamp} -> Block{minecraft:redstone_lamp}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,56, y=4,00, z=394,01]
C|1SpongeLocatableBlock{blockState=minecraft:lit_redstone_lamp, position=(-75, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@3b08a563}
C|2WorldServer{LevelName=world, DimensionId=0, DimensionType=minecraft:overworld}
```

### With Repeater:

```
ChangeBlockEvent$Break$Impl
Tra: @-70:392 Block{minecraft:redstone_torch} -> Block{minecraft:air}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
C|1EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]


ChangeBlockEvent$Modify$Impl
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-71:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
C|1SpongeLocatableBlock{blockState=minecraft:air, position=(-70, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@132b77f6}
C|2EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]


ChangeBlockEvent$Modify$Impl
Tra: @-72:392 Block{minecraft:powered_repeater} -> Block{minecraft:unpowered_repeater}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
Tra: @-73:392 Block{minecraft:redstone_wire} -> Block{minecraft:redstone_wire}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
C|1SpongeLocatableBlock{blockState=minecraft:powered_repeater[delay=1,facing=east,locked=false], position=(-72, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@197275ae}
C|2WorldServer{LevelName=world, DimensionId=0, DimensionType=minecraft:overworld}


ChangeBlockEvent$Modify$Impl
Tra: @-74:392 Block{minecraft:lit_redstone_lamp} -> Block{minecraft:redstone_lamp}
CTX|Spawn Type: SpongeSpawnType{id=sponge:dropped_item, name=DroppedItem}
CTX|Owner: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
CTX|Notifier: EntityPlayerMP['Faithcaio'/56, l='world', x=-70,70, y=4,00, z=394,69]
C|1SpongeLocatableBlock{blockState=minecraft:lit_redstone_lamp, position=(-74, 4, 392), worldId=5cb5c3ef-3732-4ffa-afc8-2c0ac429d332, worldReference=java.lang.ref.WeakReference@22351dd5}
C|2WorldServer{LevelName=world, DimensionId=0, DimensionType=minecraft:overworld}
```

</p></details>

 - [ ] Linkage Error on `Cause#toString` when cause contains an Event. (tested with ChangeBlockEvent.Post)
- [ ] Leaf Decay
    -  has no Owner
    - SpawnType:BlockSpawning is always present.
    - same LocatableBlock cause is present twice
    - Leaves Decay context somehow leaks to other events. (saw it in Player-Place/Break and Sheep-Place)

 - [ ] Tree Grow is 2 Events Break (Sapling Block) and Place for the Rest.
    -  .Grow is not fired, alternative adding a grow context?

# Teleport

- [ ] MoveEntityEvent.Teleport 
    - set to Type PLUGIN on setLocation; (except when already set by plugin?)
    - set to Type ENTITY_TELEPORT for Enderman
    - ChorusFruit(currently only has SpongeModGame in cause (add EntityLivingBase to cause in onItemUseFinish?)
    - Enderpearl (only the entity is in cause)
    - set to Type COMMAND in commands
    - dismountEnitiy?
    - change spectator
    - end portal (only back to overworld) is missing causes



<details><summary>Listener Code</summary><p>

 ### Listener Code
```
        System.out.print(event.getClass().getSimpleName() + "\n");
        for (EventContextKey<?> key : event.getContext().keys()) {
            Object o = event.getContext().get(key).get();
            System.out.print("CTX|" + key.getName() + ": " + o + "\n");
        }

        int i = 1;
        for (Object o : event.getCause().all()) {
            if (!(o instanceof Event)) {
                System.out.print("C|" + i++ + o + "\n");
            }
        }
        System.out.print("\n\n");
```

</p></details>

File: src/main/java/org/spongepowered/common/event/SpongeCauseStackManager.java [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+            int i = 0;
+            for (CauseStackFrameImpl f : this.frames) {
+                if (f == oldFrame) {
+                    offset = i;
+                    break;
+                }
+                i++;
+            }
+            final PrettyPrinter printer = new PrettyPrinter(100).add("Cause Stack Frame Corruption!").centre().hr()
+                .add("Found %n frames left on the stack. Clearing them all.", offset + 1);
```
Comment (Faithcaio) [inline src/main/java/org/spongepowered/common/event/SpongeCauseStackManager.java]: `Found %n frames left on the stack. Clearing them all.` %n is never set

File: src/main/java/org/spongepowered/common/event/SpongeCommonEventFactory.java [hunk 5]
```diff
@@ -339,32 +342,37 @@ public static CollideEntityEvent callCollideEntityEvent(net.minecraft.world.Worl
...
-        Cause.Builder builder = Cause.source(source);
-        builder.named(NamedCause.of(NamedCause.LIQUID_FLOW, worldIn));
-        if (owner != null) {
-            builder.owner(owner);
-        }
-        if (notifier != null) {
-            builder.notifier(notifier);
+        try (CauseStackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+            Sponge.getCauseStackManager().pushCause(source);
+            Sponge.getCauseStackManager().addContext(EventContextKeys.LIQUID_MIX, (World) worldIn);
```
Comment (Faithcaio) [inline src/main/java/org/spongepowered/common/event/SpongeCommonEventFactory.java]: add a `LIQUID_BREAK` or keep the previous `LIQUID_FLOW`? Nothing is mixing here

File: src/main/java/org/spongepowered/common/event/tracking/phase/packet/PlaceBlockPacketState.java [hunk 6]
```diff
@@ -93,7 +89,7 @@ public void handleBlockChangeWithUser(@Nullable BlockChange blockChange, Transac
 
     @Override
     public void associateBlockEventNotifier(PhaseContext context, IMixinWorldServer mixinWorldServer, BlockPos pos, IMixinBlockEventData blockEvent) {
-        final Player player = context.first(Player.class).get();
+        final Player player = Sponge.getCauseStackManager().getCurrentCause().first(Player.class).get();
```
Comment (Faithcaio) [inline src/main/java/org/spongepowered/common/event/tracking/phase/packet/PlaceBlockPacketState.java:92]: Player is not present when opening inventories

**Pull request merged**